onnx/defs/math/defs.cc

```diff
@@ -2006,6 +2007,301 @@ ONNX_OPERATOR_SET_SCHEMA(
+                                 "3 dimensional input with [N][n_blocks_per_col][blob_size] shape.");
+          }
+          if (b_shape.dim_size() == 3) {
+            if (b_shape.dim(1).has_dim_value() && b_shape.dim(2).has_dim_value() &&
+                !(b_shape.dim(1).dim_value() == n_blocks_per_col && b_shape.dim(2).dim_value() == blob_size)) {
+              fail_shape_inference("Input B dimensions is incompatible with the MatMulNBits specification expected "
+                                   "input with [N][n_blocks_per_col][blob_size] shape.");
+            }
+          }
+
+          if (scales_shape.dim(0).has_dim_value() && ((N * n_blocks_per_col) != scales_shape.dim(0).dim_value())) {
+            fail_shape_inference("Input scales dimensions is incompatible for MatMulNBits expected ",
+                                 "[N * n_blocks_per_col] which is [", (N * n_blocks_per_col), "] got [",
+                                  scales_shape.dim(0).dim_value(), "].");
+          }
+
+          if (ctx.hasInput(3) && hasInputShape(ctx, 3)) {  // has zero_points
+            auto zero_points_type = ctx.getInputType(3);
+            if (nullptr == zero_points_type ||
+                zero_points_type->value_case() != ONNX_NAMESPACE::TypeProto::kTensorType) {
+              fail_type_inference("inputs zero_points is expected to have a tensor type.");
+            }
+
+            const auto zero_points_shape = zero_points_type->tensor_type().shape();
+            if (zero_points_shape.dim_size() != 1) {
+              fail_type_inference("Input zero_points tensor is of wrong rank.");
+            }
+            if (zero_points_shape.dim(0).has_dim_value()) {
+              int64_t bits_per_zero_points = N * ceil((n_blocks_per_col * bits) / 8.0);
+              int64_t b_dtype = b_type->tensor_type().elem_type();
+              int64_t scales_dtype = scales_type->tensor_type().elem_type();
+              int64_t zero_points_dtype = zero_points_type->tensor_type().elem_type();
+              if (zero_points_dtype == scales_dtype &&
+                  zero_points_shape.dim(0).dim_value() != (N * n_blocks_per_col)) {
+                fail_shape_inference("Input zero_points dimensions is incompatible for MatMulNBits expected ",
+                                     "[N * n_blocks_per_col] which is [", (N * n_blocks_per_col), "] got [",
+                                     zero_points_shape.dim(0).dim_value(), "].");
+              }
+              if (zero_points_dtype == b_dtype && zero_points_shape.dim(0).dim_value() != bits_per_zero_points) {
+                fail_shape_inference("Input uint8 zero_points dimensions is incompatible for MatMulNBits expected ",
+                                     "[N * CeilDiv(n_blocks_per_col * bits, 8)] which is [", bits_per_zero_points,
+                                     "] got [", zero_points_shape.dim(0).dim_value(), "].");
+              }
```

Comment:
I think a check that `(zero_points_dtype == scales_dtype) || (zero_points_dtype == b_dtype)` could be added. I assume that is required? Eg., is it allowed for T1 to be float and T3 to be float16? 

Also: related to Liqun's point above: if we allow scale-type to be different, will need to change the conditions accordingly here.